pkg/collector/corechecks/cluster/orchestrator/manifest_buffer_test.go

```diff
@@ -0,0 +1,171 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-present Datadog, Inc.
+
+//go:build kubeapiserver && orchestrator
+// +build kubeapiserver,orchestrator
+
+package orchestrator
+
+import (
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/require"
+	"gotest.tools/assert"
+
+	model "github.com/DataDog/agent-payload/v5/process"
+
+	"github.com/DataDog/datadog-agent/pkg/aggregator"
+	"github.com/DataDog/datadog-agent/pkg/collector/check"
+	"github.com/DataDog/datadog-agent/pkg/metrics"
+	"github.com/DataDog/datadog-agent/pkg/serializer"
+)
+
+// TestOrchestratorManifestBuffer tests manifest buffer
+// This test sets maximum beffer size to 2 and we have 5 manifests in total
+// The maximum manifests per message is 100 by default
+// Therefore we should have 3 messages to send
+// message 1 contains manifest 1 and 2
+// message 2 contains manifest 3 and 4
+// message 3 contains manifest 5
+func TestOrchestratorManifestBuffer(t *testing.T) {
+	orchCheck := OrchestratorFactory().(*OrchestratorCheck)
+	mb := NewManifestBuffer(orchCheck)
+	mb.Cfg.MaxBufferedManifests = 2
+	mb.Cfg.ManifestBufferFlushInterval = 3 * time.Second
+
+	noopSender := noopSender{}
+	mb.Start(&noopSender)
+
+	// Send 5 manifests to the buffer
+	for i := 1; i <= 5; i++ {
+		mb.ManifestChan <- &model.Manifest{
+			Type: int32(i),
+		}
+	}
+
+	mb.Stop()
+
+	// The buffer should be flushed 3 times
+	require.Len(t, noopSender.OrchestratorManifestOut, 3)
+
+	i := int32(1)
+	for _, collectorManifests := range noopSender.OrchestratorManifestOut {
```

Comment:
Thanks for the test!
I think it looks great already. I do think that you could re-use the existing mock which could give you more "power" on how the manifest could look like during each buffer iteration. As of right now the check code (where you iterate through the out chan) is difficult to understand. 

Something like this should work (need to test for additional iterations):

```
Index: pkg/collector/corechecks/cluster/orchestrator/manifest_buffer_test.go
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/pkg/collector/corechecks/cluster/orchestrator/manifest_buffer_test.go b/pkg/collector/corechecks/cluster/orchestrator/manifest_buffer_test.go
--- a/pkg/collector/corechecks/cluster/orchestrator/manifest_buffer_test.go	(revision 6749f75c3fdc1d31be9c2c071cc4a0c0b2ae0d06)
+++ b/pkg/collector/corechecks/cluster/orchestrator/manifest_buffer_test.go	(date 1668677718294)
@@ -12,15 +12,12 @@
 	"testing"
 	"time"
 
-	"github.com/stretchr/testify/require"
-	"gotest.tools/assert"
-
 	model "github.com/DataDog/agent-payload/v5/process"
-
-	"github.com/DataDog/datadog-agent/pkg/aggregator"
+	"github.com/DataDog/datadog-agent/pkg/aggregator/mocksender"
 	"github.com/DataDog/datadog-agent/pkg/collector/check"
-	"github.com/DataDog/datadog-agent/pkg/metrics"
 	"github.com/DataDog/datadog-agent/pkg/serializer"
+	"github.com/stretchr/testify/mock"
+	"go.uber.org/atomic"
 )
 
 // TestOrchestratorManifestBuffer tests manifest buffer
@@ -35,133 +32,25 @@
 	mb := NewManifestBuffer(orchCheck)
 	mb.Cfg.MaxBufferedManifests = 2
 	mb.Cfg.ManifestBufferFlushInterval = 3 * time.Second
+	mb.Cfg.MsgGroupRef = atomic.NewInt32(1)
+	a := mocksender.NewMockSender(check.ID(rune(1)))
+	a.On("OrchestratorManifest", mock.Anything, mock.Anything).Return()
 
-	noopSender := noopSender{}
-	mb.Start(&noopSender)
+	mb.Start(a)
+	b := &model.Manifest{
+		Type: int32(1),
+	}
 
 	// Send 5 manifests to the buffer
-	for i := 1; i <= 5; i++ {
-		mb.ManifestChan <- &model.Manifest{
-			Type: int32(i),
-		}
+	mb.ManifestChan <- b
+
+	c := model.CollectorManifest{
+		GroupId:   2,
+		GroupSize: 1,
+		Manifests: []*model.Manifest{b},
 	}
 
 	mb.Stop()
-
-	// The buffer should be flushed 3 times
-	require.Len(t, noopSender.OrchestratorManifestOut, 3)
-
-	i := int32(1)
-	for _, collectorManifests := range noopSender.OrchestratorManifestOut {
-		// The maximum manifests per message is 100 by default, therefore we should only have one message created by each flush
-		require.Len(t, collectorManifests, 1)
-		collectorManifest := collectorManifests[0].(*model.CollectorManifest)
-		for _, m := range collectorManifest.Manifests {
-			assert.Equal(t, i, m.Type)
-			i++
-		}
-	}
-
-}


```


Feel free to experiment and if you don't think it is worth it we can revisit with the approach you have started. 
If we want to go ahead with your current approach i suggest to change the code to test the chan at each iteration step and refactor the code around the part `The maximum manifests per message` as I don't fully follow the comment

Reply:
Thanks for the review ! I didn't realise we have already mockSender. It was so painful to implement a noopSender.
I addressed your feedback in [67b2ae0](https://github.com/DataDog/datadog-agent/pull/13943/commits/67b2ae062dae3c690500306fc73feea419829fb5)

Reply:
I just realized that I missed that in the patch :D 

```
	a.Mock.AssertCalled(t, "OrchestratorManifest", []serializer.ProcessMessageBody{&c}, "")
```

But I like your approach as well.

The only thing imo missing would be something like this, but thats my style preference and the test as it is should suffice

```
mb.ManifestChan <-  a
assertLen(globalManifestBuffer, 1)
mb.ManifestChan <-  a
assertLen(globalManifestBuffer, 2)

// flush
mb.ManifestChan <-  a
assertLen(globalManifestBuffer, 1)
```